IR Interpreter

hrydgard opened the pull request:
So, this is different from the previous one that used MIPS instructions as an IR directly.

This translates code into a fast-to-interpret and easy-to-analyze IR (intermediate representation). Gets rid of delay slots and other MIPS silliness (though there's a small wart with "likely" branch instructions) while still keeping the MIPS register set and semantics, mostly. It then optimizes and caches these blocks like the real JITs and executes them.

This new IR should also suitable as a first step for things like more proper full-function compilation, SSA, etc. you'd probably want to first use this translator, run the constant propagation pass, then link multiple of these blocks together and lift into an AST, transform that into SSA, perform more optimizations and then do register allocation and code generation. Or something simpler :)

Anyhow, this is around 50% faster than the current interpreter already, so will be useful for non-JIT platforms like iOS, even if no more advanced compilation is done.

It's completely missing things like CPU floating point rounding mode for now.

hrydgard committed:
Initial commit for IRJit

hrydgard committed:
It's kind of starting to run

hrydgard committed:
Enough to run cpu_alu.prx.

hrydgard committed:
Fix a performance issue in CallSyscall

hrydgard committed:
Disable debug file logging, fix issue with replacement functions, etc

hrydgard committed:
Initial work on mult and friends. Fix another bug.

hrydgard committed:
ir-jit: Get rid of the regcache. Should be replaced with optimization passes.

hrydgard committed:
Use the regcache in a new (incomplete) pass, PropagateConstants.

hrydgard committed:
Make the IRJit core selectable in developer tools

hrydgard committed:
Add support for float store/load

hrydgard committed:
Fix wsbw

hrydgard committed:
ir-jit: Fix bug in ext

hrydgard committed:
ir-jit: Fix bugs in rounding

hrydgard committed:
ir-jit: Re-enable mult. Fix bvf/bvt

hrydgard committed:
Fix IRTEMP clash bug. Add more cases to the constant propagation pass.

hrydgard committed:
More simplify pass

hrydgard committed:
Initial VFPU

unknownbrackets commented on Core/MIPS/IR/IRJit.cpp:
```diff
@@ -0,0 +1,430 @@
+
+		// Let's try that one more time.  We won't get back here because we toggled the value.
+		js.startDefaultPrefix = false;
+		// TODO ARM64: This crashes.
```
Does it still?  I think we want this, both here and in arm64jit.

-[Unknown]

hrydgard replied:
Hmm, yeah, we do. Have not tried recently on ARM64.

unknownbrackets commented on Core/MIPS/IR/IRCompALU.cpp:
```diff
@@ -0,0 +1,381 @@
+	MIPSGPReg rd = _RD;
+	MIPSGPReg rt = _RT;
+	MIPSGPReg rs = _RS;
+	// Not sure if ARM64 wraps like this so let's do it for it.  (TODO: According to the ARM ARM, it will indeed mask for us so this is not necessary)
```
Should remove this comment.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRCompALU.cpp:
```diff
@@ -0,0 +1,381 @@
+
+	case 0x4: //ins
+	{
+		logBlocks = 1;
```
Debugging.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRCompALU.cpp:
```diff
@@ -0,0 +1,381 @@
+		ir.Write(IROp::Ext16to32, rd, rt);
+		break;
+
+	case 20: //bitrev
```
We ought to IR this, no?  ARM64 has RBIT.

-[Unknown]

hrydgard replied:
Right, but it's a rare instruction so I just left it for later.

unknownbrackets commented on Core/MIPS/IR/IRCompALU.cpp:
```diff
@@ -0,0 +1,381 @@
+		break;
+
+	case 26: //div
+		DISABLE;
```
Should probably remove these DISABLEs, unless they're not working?  Maybe a comment if so?

-[Unknown]

hrydgard replied:
They aren't, so not yet, heh. Just did Mult as it's by far the most common.

unknownbrackets commented on Core/MIPS/IR/IRCompBranch.cpp:
```diff
@@ -0,0 +1,379 @@
+	MIPSOpcode delaySlotOp = GetOffsetInstruction(1);
+	bool delaySlotIsNice = IsDelaySlotNiceReg(op, delaySlotOp, rt, rs);
+
+	ir.Write(IROp::Downcount, 0, js.downcountAmount & 0xFF, js.downcountAmount >> 8);
```
Hmm, I think this won't account for the delay slot, right?

-[Unknown]

hrydgard replied:
Adding 1 to them all.

unknownbrackets replied:
Well, ideally we should add `MIPSInstructionEstimate(delaySlotOp)` or whatever it was, right?

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRCompBranch.cpp:
```diff
@@ -0,0 +1,379 @@
+
+	MIPSGPReg lhs = rs;
+	MIPSGPReg rhs = rt;
+	if (!delaySlotIsNice) {  // if likely, we don't need this
```
Just checking, realize this isn't needed atm - we should do `!likely &&` right?

Also, we could fairly easily optimize out only `rs` or only `rt` if only one of them are written.  Can be later.

-[Unknown]

hrydgard replied:
Right, fixing.

Indeed, though a small win.

unknownbrackets commented on Core/MIPS/IR/IRCompBranch.cpp:
```diff
@@ -0,0 +1,379 @@
+	if (!likely)
+		CompileDelaySlot();
+
+	FlushAll();
```
Hmm, so I guess these flushalls (at this point) don't actually do anything, right?  Otherwise we would somehow want to flush all EXCEPT IRTEMP_LHS/IRTEMP_RLHS.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRCompBranch.cpp:
```diff
@@ -0,0 +1,379 @@
+	// The CC flags here should be opposite of the actual branch becuase they skip the branching action.
+	switch (op >> 26) {
+	case 4: BranchRSRTComp(op, IRComparison::NotEqual, false); break;//beq
+	case 5: BranchRSRTComp(op, IRComparison::Equal,  false); break;//bne
```
Extra space on a couple these (ubernit.)

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRCompBranch.cpp:
```diff
@@ -0,0 +1,379 @@
+	// The behavior is undefined - the CPU may take the second branch even if the first one passes.
+	// However, it does consistently try each branch, which these games seem to expect.
+	bool delaySlotIsBranch = MIPSCodeUtils::IsVFPUBranch(delaySlotOp);
+	if (!likely)
```
`&& !delaySlotIsBranch`, right?  That's the point of going to + 4 in that case.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRCompBranch.cpp:
```diff
@@ -0,0 +1,379 @@
+	// If we're in a delay slot, this is off by one.
+	const int offset = js.inDelaySlot ? -1 : 0;
+	RestoreRoundingMode();
+	js.downcountAmount = -offset;
```
This downcount is never written, right?

-[Unknown]

hrydgard replied:
Nice catch!

unknownbrackets commented on Core/MIPS/IR/IRCompFPU.cpp:
```diff
@@ -0,0 +1,246 @@
+
+	switch (op >> 26) {
+	case 49: //FI(ft) = Memory::Read_U32(addr); break; //lwc1
+	{
```
Can probably drop the curlies here to make it purdier.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRCompFPU.cpp:
```diff
@@ -0,0 +1,246 @@
+}
+
+void IRJit::Comp_FPUComp(MIPSOpcode op) {
+	DISABLE;
```
Was this temp debugging or is it still broken?  Comment?

-[Unknown]

hrydgard replied:
Not implemented yet

unknownbrackets commented on Core/MIPS/IR/IRCompBranch.cpp:
```diff
@@ -0,0 +1,379 @@
+	using namespace Arm64Gen;
+
+void IRJit::BranchRSRTComp(MIPSOpcode op, IRComparison cc, bool likely)
+{
```
nit: Curly pos.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRCompFPU.cpp:
```diff
@@ -0,0 +1,246 @@
+}
+
+void IRJit::Comp_mxc1(MIPSOpcode op)
+{
```
nit: Curly pos.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRCompFPU.cpp:
```diff
@@ -0,0 +1,246 @@
+		if (fs == 31) {
+			DISABLE;
+		}
+		else if (fs == 0) {
```
nit: Curly pos.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRCompFPU.cpp:
```diff
@@ -0,0 +1,246 @@
+			return;
+		}
+		if (fs == 31) {
+			DISABLE;
```
Probably should just give this its own dedicated IR op, right?

I guess if we don't disable here, though, we *have* to interpret.

-[Unknown]

hrydgard replied:
Adding TODO

unknownbrackets commented on Core/MIPS/IR/IRCompLoadStore.cpp:
```diff
@@ -0,0 +1,161 @@
+
+namespace MIPSComp {
+	void IRJit::Comp_ITypeMemLR(MIPSOpcode op, bool load) {
+		DISABLE;
```
I didn't realize until now that arm64jit still has this disabled.  Wonder what's wrong...

Do you think we should do the lwl/lwr combine in IR, or leave it for a later pass?

-[Unknown]

hrydgard replied:
That's a very good question that I haven't come up with an answer to yet :) Ideally we would be able to emit the full algorithm with shifts and all, and then have the optimizers shake it out when possible, though that may be a bit utopian...

unknownbrackets commented on Core/MIPS/IR/IRCompBranch.cpp:
```diff
@@ -0,0 +1,379 @@
+
+	FlushAll();
+	ir.Write(ComparisonToExit(cc), ir.AddConstant(GetCompilerPC() + 8), lhs, rhs);
+	// This makes the block "impure" :(
```
If we wanted to keep the block pure, I guess an option would be to create another dedicated block for the likely delay slot, and then have it jump to the exit in that block.  Would take some work to hook up.  We'd also need to mark it as dependent for invalidation.

Then all blocks could be pure, and we'd just need an optimization pass to combine short blocks or something.

-[Unknown]

hrydgard replied:
Yeah, I thought about that too, but decided to hold off until I actually do multi-block things. A problem is that these new blocks would have no unique address that can be used reliably.

unknownbrackets replied:
We have that in our proxy blocks now, too.  It's a bit complicated but I think it does work... we don't ever really need to get a block by address, it's just invalidation.

-[Unknown]

hrydgard replied:
Right .. it should be solvable. Not for the first iteration of this though :)

unknownbrackets commented on Core/MIPS/IR/IRInst.cpp:
```diff
@@ -0,0 +1,589 @@
+	}
+}
+
+u32 IRInterpret(MIPSState *mips, const IRInst *inst, const u32 *constPool, int count) {
```
Assuming we add VFPU, etc. this will get a lot longer.  What do you think about having it in a dedicated file?

-[Unknown]

hrydgard replied:
Yeah, starting to get to that point.

unknownbrackets commented on Core/MIPS/IR/IRInst.cpp:
```diff
@@ -0,0 +1,589 @@
+		}
+
+		case IROp::Break:
+			Crash();
```
Definitely want this one to be optional, at least when fastmem is off.

-[Unknown]

hrydgard replied:
Not 100% sure exactly how it should be implemented though..

unknownbrackets replied:
If you have ignore bad addresses on, it just keeps on trucking.  This has been enough to make some games work in the past.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRInst.cpp:
```diff
@@ -0,0 +1,589 @@
+		default:
+			Crash();
+		}
+#ifdef _DEBUG
```
I guess we probably won't want this long-term for debugging, right?

-[Unknown]

hrydgard replied:
No, but it's been useful. It'll go away long term :)

unknownbrackets commented on Core/MemMapFunctions.cpp:
```diff
@@ -87,7 +87,7 @@ inline void ReadFromHardware(T &var, const u32 address) {
 	} else {
 		// In jit, we only flush PC when bIgnoreBadMemAccess is off.
-		if (g_Config.bJit && g_Config.bIgnoreBadMemAccess) {
+		if (g_Config.iCpuCore != CPU_CORE_INTERPRETER && g_Config.bIgnoreBadMemAccess) {
```
I think this can be only when in jit.

Really, this is just so that lw/sw/etc. will not log a completely wrong PC and confuse people.  We don't flush the PC in slowmem when calling these funcs.

We do flush before syscalls, and we only use unchecked in IR interpret.  I wonder what this would be set to in a IR-based jit...

Ideally we'd still log the address in syscalls anyway, maybe it'd be better to have a separate func...

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRInst.cpp:
```diff
@@ -0,0 +1,589 @@
+	case IRTEMP_1: return "irtemp1";
+	case IRTEMP_LHS: return "irtemp_lhs";
+	case IRTEMP_RHS: return "irtemp_rhs";
+	default: return "(unk)";
```
Do hi/lo/fpcond/etc. count?

-[Unknown]

hrydgard replied:
I'm not sure we really want to access those through this ugly mechanisms, they probably deserve dedicated instructions.

unknownbrackets commented on Core/MIPS/IR/IRInst.cpp:
```diff
@@ -0,0 +1,589 @@
+		snprintf(buf, bufSize, "%s", GetGPRName(param));
+		break;
+	case 'F':
+		snprintf(buf, bufSize, "r%d", param);
```
Hmm, not `f`?  Or I guess even `currentDebugMIPS->GetRegName(1, param)`?

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRInst.cpp:
```diff
@@ -0,0 +1,589 @@
+	if (meta->types[1] && meta->types[0] != '_') {
+		strcat(bufDst, ", ");
+	}
+	if (meta->types[2] && meta->types[1] != '_') {
```
Hmm.  `Break` has `""` as the types, so could `types[1]` be something unexpected?  Or would it always be zero?  I would hope it's always zero...

-[Unknown]

hrydgard replied:
it's defined as a 4 char array ... so I hope the rest get 0'd. Not sure though..

unknownbrackets commented on Core/MIPS/IR/IRInst.h:
```diff
@@ -0,0 +1,277 @@
+	Bad,
+};
+
+// Hm, unused
```
Could be used for a theoretical loop unrolling pass.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRInst.h:
```diff
@@ -0,0 +1,277 @@
+
+enum class IRParam {
+	Ignore = '_',
+	UImm8 = 'U',
```
Hmm, I think you're using `I` in practice, right?

-[Unknown]

hrydgard replied:
Uh, yeah. Oops. Never ended up using this enum at all, gone.

unknownbrackets commented on Core/MIPS/IR/IRJit.cpp:
```diff
@@ -0,0 +1,430 @@
+	js.startDefaultPrefix = mips_->HasDefaultPrefix();
+	js.currentRoundingFunc = convertS0ToSCRATCH1[0];
+	u32 size = 128 * 1024;
+	blTrampolines_ = kernelMemory.Alloc(size, true, "trampoline");
```
Hmm, I think this isn't used, I see where you're going here.  Would take careful invalidation handling...

Maybe we should comment out / TODO as long as this memory isn't being used?

-[Unknown]

hrydgard replied:
Oops, yeah that was a left behind side track.

unknownbrackets commented on Core/MIPS/IR/IRJit.cpp:
```diff
@@ -0,0 +1,430 @@
+	js.currentRoundingFunc = convertS0ToSCRATCH1[0];
+	u32 size = 128 * 1024;
+	blTrampolines_ = kernelMemory.Alloc(size, true, "trampoline");
+	logBlocks = 12;
```
Debugging.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRJit.cpp:
```diff
@@ -0,0 +1,430 @@
+}
+
+void IRJit::ClearCache() {
+	ILOG("ARM64Jit: Clearing the cache!");
```
The comment is out of date.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRJit.cpp:
```diff
@@ -0,0 +1,430 @@
+				mips_->pc = IRInterpret(mips_, block->GetInstructions(), block->GetConstants(), block->GetNumInstructions());
+			} else {
+				if (mips_->pc == 0x0880de94)
+					logBlocks = 10;
```
Debugging.

-[Unknown]

hrydgard replied:
Oops

unknownbrackets commented on Core/MIPS/IR/IRJit.cpp:
```diff
@@ -0,0 +1,430 @@
+	js.PrefixStart();
+	ir.Clear();
+
+	int partialFlushOffset = 0;
```
Can kill this, I think.  Should probably kill it in arm64jit too...

-[Unknown]

hrydgard replied:
Done

unknownbrackets commented on Core/MIPS/IR/IRJit.cpp:
```diff
@@ -0,0 +1,430 @@
+		js.compilerPC += 4;
+		js.numInstructions++;
+
+		if (ir.GetConstants().size() > 64) {
```
Maybe we should make `64` a const up higher.  I wonder if there'd be any measurable gain to using 128... probably not.

-[Unknown]

hrydgard replied:
Just did it as a temporary workaround for Tekken hitting 256 after constant propagation when I had 128.  Not sure how it managed to do that though...

unknownbrackets replied:
Oh, wow.

Actually, I think that's something that happens in our current jit, and really could be optimized better.  Separating it like this really gives us the opportunity to improve.

Basically, what can happen is this:

```
lui t0, 0x1234
ori t0, t0, 0x5678

lw a0, 0(t0)
lw a0, 4(t0)
lw a0, 8(t0)
lw a0, 12(t0)
lw a0, 16(t0)
lw a0, 20(t0)
...
lw a0, 252(t0)
```

`PropagateConstants` would currently generate a ton of constants for this, and if 0/4/8/etc. were reused often with different bases, it could explode.  This doesn't happen in jit, but other scenarios do, where we could more optimally hold on to a "common base" and end up emitting less constant imms in total.

-[Unknown]

hrydgard replied:
Hm, yes. But I think it's best to actually fully evaluate the constants, and then "optimize backwards" to find a common base, that way we won't miss some cases if multiple base registers are used. It does increase the risk of hitting the new "constant ceiling" though...

unknownbrackets commented on Core/MIPS/IR/IRJit.h:
```diff
@@ -0,0 +1,297 @@
+	void InvalidateICache(u32 addess, u32 length);
+	int GetNumBlocks() const { return (int)blocks_.size(); }
+	int AllocateBlock(int emAddr) {
+		blocks_.emplace_back(IRBlock(emAddr));
```
I bet Symbian will hate this.

-[Unknown]

hrydgard replied:
Probably, heh.

unknownbrackets commented on Core/MIPS/IR/IRPassSimplify.cpp:
```diff
@@ -0,0 +1,199 @@
+void SimplifyInPlace(IRInst *inst, int count, const u32 *constPool) {
+	for (int i = 0; i < count; i++) {
+		switch (inst[i].op) {
+		case IROp::AddConst:
```
I think this is just an example, but can probably add `OrConst` here too.

-[Unknown]

hrydgard replied:
SimplifyInPlace can probably be removed now that I have the constantpropagation pass which does similar simplifications, too..

unknownbrackets replied:
Right.  Although, I suppose some in-place optimizations (like combining, by nop'ing) could be done, it probably belongs in other passes.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRRegCache.cpp:
```diff
@@ -0,0 +1,61 @@
+
+void IRRegCache::FlushAll() {
+	for (int i = 0; i < TOTAL_MAPPABLE_MIPSREGS; i++) {
+		//if (i < IRTEMP_0)
```
Those can be constants too right?  I think we can just remove this line.

-[Unknown]

hrydgard replied:
Yes, that was a misguided move. I found, though, that after constant propagation some dead SetConstants to IRTEMP sometimes remain. Need another sort of pass to get rid of those.

unknownbrackets replied:
Ah, I know that some mips will even generate dead reg stores, e.g.:

```
li a0, 5
add a0, a1, a2
```
(I don't remember digging in, maybe the cases I've seen have all been branch things...)

We could probably have a generic pass to eliminate all clobbered / unused values.

-[Unknown]

hrydgard replied:
Clobbered values like this do actually get eliminated by this constant propagation pass (MapDirtyIn etc).

unknownbrackets replied:
Ah, right.

-[Unknown]

unknownbrackets commented on Core/MIPS/JitCommon/JitCommon.cpp:
```diff
@@ -48,7 +50,7 @@ namespace MIPSComp {
 		return new MIPSComp::ArmJit(mips);
 #elif defined(ARM64)
-		return new MIPSComp::Arm64Jit(mips);
+		return new MIPSComp::IRJit(mips);
```
Debugging, right?

-[Unknown]

hrydgard replied:
More like strange copy paste mistake of some sort. Thanks.

unknownbrackets commented on Core/MIPS/x86/Asm.cpp:
```diff
@@ -40,7 +40,7 @@ namespace MIPSComp
 //TODO - make an option
 //#if _DEBUG
-static bool enableDebug = false; 
+	static bool enableDebug = true;
```
Debugging.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRPassSimplify.cpp:
```diff
@@ -0,0 +1,199 @@
+	for (int i = 0; i < (int)in.GetInstructions().size(); i++) {
+		IRInst inst = in.GetInstructions()[i];
+		bool symmetric = true;
+		switch (inst.op) {
```
We should do something like `if (out.GetContants.size() > 128) goto doDefault;`...

-[Unknown]

hrydgard committed:
Address a bunch of comments, thanks for the review

unknownbrackets commented on Core/MIPS/IR/IRCompBranch.cpp:
```diff
@@ -363,6 +368,9 @@ void IRJit::Comp_Syscall(MIPSOpcode op) {
 	RestoreRoundingMode();
 	js.downcountAmount = -offset;
 
+	int dcAmount = js.downcountAmount + 1;
```
In this case there's no delay slot, and if we're already in a delay slot, it should be -1.  I think the other jits do this.

-[Unknown]

unknownbrackets commented on Core/MemMapFunctions.cpp:
```diff
@@ -123,7 +123,7 @@ inline void WriteToHardware(u32 address, const T data) {
 	} else {
 		// In jit, we only flush PC when bIgnoreBadMemAccess is off.
-		if (g_Config.bJit && g_Config.bIgnoreBadMemAccess) {
+		if (g_Config.iCpuCore != CPU_CORE_INTERPRETER && g_Config.bIgnoreBadMemAccess) {
```
I guess might as well change this one too, heh.  Sorry.

-[Unknown]

hrydgard replied:
Heh, thanks.

hrydgard committed:
Move the IR interpreter out into its own file. Rename it in the UI.  First CMake and Android fixes.

hrydgard committed:
Additional fixes and buildfixes

hrydgard committed:
Another buildfix

hrydgard committed:
Fix the fix

hrydgard committed:
jit-ir: Less instructions cause flushing in constant propagation.

unknownbrackets committed:
Qt, gcc, and Symbian buildfixes.

hrydgard committed:
More constant propagation

unknownbrackets commented on Core/MIPS/IR/IRRegCache.h:
```diff
@@ -0,0 +1,47 @@
+	IRRegCache(IRWriter *ir);
+
+	void SetImm(int r, u32 immVal) {
+		reg_[r].isImm = true;
```
We should add an `assert(r != MIPS_REG_ZERO);` or similar here.

-[Unknown]

unknownbrackets commented on Core/MIPS/IR/IRPassSimplify.cpp:
```diff
@@ -0,0 +1,273 @@
+		default:
+		{
+		doDefaultAndFlush:
+			gpr.FlushAll();
```
Hmm.  In the case of `default`, we could probably just only flush `dest, src1, src2` based on IRMeta, right?  It's only for exits, syscall, interpret, and replacements that we need to flushall.  We might need to flush certain special regs for a few others though, so I guess this is the safest default...

-[Unknown]

hrydgard replied:
Absolutely - and yes, we do need to flush it all before running those particular ops. But yeah, just a safe default for now.

unknownbrackets replied:
Maybe we can have a temp that discards temps (like IRTEMP0) that are no longer read from due to other optimization passes (like const folding.)

-[Unknown]

unknownbrackets committed:
jit-ir: Skip const flush on downcount op.

This allows discarding more unused constants.

hrydgard committed:
Skip const flush on set float constant

unknownbrackets committed:
jit-ir: Add a utility func for applying passes.

unknownbrackets committed:
jit-ir: Propagate single-op constants.

unknownbrackets committed:
jit-ir: Skip flushing when updating PC.

unknownbrackets committed:
jit-ir: Propagate constants even for overlaps.

unknownbrackets committed:
jit-ir: Propagate constants for shifts.

This might optimize away an IRTEMP_0 in such cases.

unknownbrackets committed:
jit-ir: Flush consts better for a few f/v ops.

unknownbrackets committed:
jit-ir: Fix SetConstV.

Oops.

unknownbrackets committed:
jit-ir: Add the rest to PropagateConstants.

Just for completeness.

unknownbrackets commented on Core/MIPS/IR/IRCompVFPU.cpp:
```diff
@@ -0,0 +1,388 @@
+		switch (op >> 26) {
+		case 54: //lv.q
+		{
+			// TODO: Add vector load/store instruction to the IR
```
Hmm, how should we do this?  `vregs` may (due to swizzle) contain all sorts of disparate regs.  In general, this is gonna be an issue for all the vfpu ops.  Tricky.

Maybe we could have a LoadFloatVx4 only when vregs are in order, since that's really the case to optimize?  In general SIMD seems tricky...

We're at 109 IR ops now, there are about 56 total VFPU ops.  I guess we could theoretically afford to have an x4 version of each for when they're in-order (which could be applied by a pass after reordering instructions, even, in case they're in column/row order but there's enough ops, or w/e.)

-[Unknown]

hrydgard replied:
Yeah, this will be the trickiest part of the whole thing. I can see a few options:

  * Take it all down to scalar, optimize, and then try to get back to SIMD in the backends that consume the IR
  * Implement already in the IR pass a similar approach that we have in the x64 backend, where we join things into SIMD ops and write back again, emitting friendly modern 4x1 SIMD ops but may be pretty far from optimal in places. That makes the job easy for the backends that consume the IR as they'll just need to map IR-simd to their own SIMD and regalloc, but may lose some opportunities.
* Emit direct analogues to the PSP's operations (with a few simplifications like eliminating the matrix ops) and let the backends do the same work they do now.

Really not sure what is best.

hrydgard replied:
Oh yeah, for the particular op lv.q and sv,q swizzle isn't supported so we don't need to worry in these particular instructions. For all others though...  I think we'll split the swizzles into a few ops into temp registers, like "selective_abs" and "shuffle" etc.

unknownbrackets replied:
Well, you still have column vs row with lv.q/sv.q.  One is SIMD friendly, the other isn't totally.

I'm thinking having an IR pass to try to find "SIMD friendly" operations (like x86) would be nice.  Yes, we may miss some opportunities there, but it doesn't mean we can't also apply approach 1 (make the backends find the optimizations) as well in those cases.

It would make implementing SIMD on ARM/ARM64 much easier, I think, to just get the bugs right once if we can, for most ops.

-[Unknown]

hrydgard replied:
One issue with option 1 is that we'll need a new set of 4x1 pseudo registers for the IR to allocate from and operate in, but that can of course be easily added. Wonder how many would make sense, so that we don't need to operate a fully featured register cache for them in the IR stage already...

unknownbrackets replied:
Well, if they're sequential, than the IR can just specify the first register in the sequence, right?

-[Unknown]

hrydgard replied:
Yeah, but we'll need temps too, for prefixes (first "prefix" into a temp register, than use it for computations). But as there are "only" 128 vector regs we do have some space for extra ones...

unknownbrackets committed:
jit-ir: Remove redundant simplify pass.

This is just doing the same thing as the const folding pass, really.

unknownbrackets committed:
jit-ir: Fix replacement hooks.

hrydgard committed:
Merge pull request #8727 from unknownbrackets/ir-jit

Apply more constant propagation to IR

unknownbrackets committed:
jit-ir: Correct downcount handling.

Oops, was wrong - already accounted for delay slots.  Clear so we don't
double count when emitting a syscall.

Fixes FF4 utility msg flickering.

hrydgard committed:
Merge pull request #8731 from unknownbrackets/ir-jit

jit-ir: Correct downcount handling

hrydgard committed:
Refactor prep: Split JitInterface into MIPSFrontendInterface and JitInterface

hrydgard committed:
Separate the IR frontend from the IR "Jit"

hrydgard committed:
Complete the separation of the IR compiler frontend from the "Jit"

hrydgard committed:
IRFrontend shouldn't know about IRBlock

hrydgard committed:
Do the voffset remapping before the IR. This will let us easily add some virtual VFPU registers for the IR to the end, plus it's slightly faster.

hrydgard committed:
More VFPU

unknownbrackets committed:
jit-ir: Optimize out unused temp regs.

This way, if constants have made the temp obsolete (common with ins, for
example), it won't even get set anymore.

unknownbrackets committed:
jit-ir: Expand unused regs to regular GPRs.

hrydgard committed:
IR: Some more VFPU

hrydgard committed:
IR: Merge the FPU and VFPU instruction sets, no reason to keep them apart

hrydgard committed:
Prefix prep

hrydgard committed:
Prefix prep

hrydgard committed:
IR: Lots more VFPU support, some with SIMD

hrydgard committed:
Crashfix in savestate (hmmmm...)

hrydgard committed:
IR: SIMD another matrix orientation. Fix various issues.

hrydgard committed:
IR: More VFPU. Support normal fp compares.

hrydgard committed:
Forgot this

hrydgard committed:
Bugfixes

hrydgard committed:
IR: vcmp, vcmov, vhdp

unknownbrackets committed:
jit-ir: Prevent reading ahead for each reg write.

unknownbrackets committed:
jit-ir: ExitToReg doesn't write to registers.

unknownbrackets commented:
Eventually we should just merge this (needs some remaining buildfixes I guess), it's mostly well-contained at least, even if there are bugs.

-[Unknown]

unknownbrackets commented:
I wonder.  For this sequence:

```
088b2f54   divu t0, a2
088b2f58   beql a2, zero, ->$088b2f60
088b2f5c   break  --- unknown ---
```

Maybe we could just pretend there was no beql at all?  It can be relatively common, and it's just skipping the break when the divisor is non zero.

-[Unknown]

unknownbrackets committed:
jit-ir: Add some missing CONDITIONAL_DISABLEs.

unknownbrackets committed:
jit-ir: Avoid flushing on a few Vec4 ops.

unknownbrackets committed:
jit-ir: Fix non-SSE Vec4Scale.

unknownbrackets committed:
jir-ir: Correct vftm SIMD regs.

unknownbrackets committed:
jit-ir: Apply prefixes for vector init ops.

Without this, Gods Eater Burst is horribly broken.

unknownbrackets committed:
jit-ir: Improve vidt to handle more cases.

unknownbrackets committed:
jit-ir: Add nan/inf compares.

Without this, Gods Eater Burst crashes before going in game.

hrydgard committed:
Merge pull request #8732 from unknownbrackets/ir-jit

Add a pass to remove temporary regs

hrydgard commented:
@unknownbrackets yes that should be safe. We could even do it before the IR to not have to create a new block, although ideally future multi-block compilation will optimize it away.

unknownbrackets committed:
jit-ir: Optimize out beql; break; sequences.

These are often used following divs, and are harmless.  Things get a bit
easier if we just never compile them.

hrydgard committed:
Merge pull request #8733 from unknownbrackets/ir-jit

jit-ir: Optimize out beql; break; sequences

hrydgard committed:
Minor build and other fixes

hrydgard committed:
Add a couple more passes (2-op, optimize f<->v fp moves)

hrydgard committed:
Optimize some common prefixes

hrydgard committed:
Simple workaround for timing issue with coreState after syscall.

Also fixes off by one in ForceCheck.

hrydgard committed:
Simplifications and fixes

hrydgard committed:
More VFPU, vmmul thoughts

unknownbrackets committed:
Qt: Remove CPU core option from menu.

We've removed on Windows too, and this fixes a build error.

unknownbrackets committed:
Hack to make Symbian build.

unknownbrackets committed:
Try enabling Travis caching.

hrydgard committed:
Merge pull request #8735 from hrydgard/ir-jit-test

Buildfixes + Travis caching

hrydgard commented:
Thanks for the fixes @unknownbrackets , I'll merge this later today after a few more changes.

Should get around to fixing the rounding as well, as this will produce broken saves in God Eater etc...

hrydgard committed:
Warning fixes
